### PR TITLE
Build tree to get all leaf nodes to abort process

### PR DESCRIPTION
**Pull request by boyan-velinov**
#### Description: 
When process is aborted, AbortProcessAction should wait
until all running subprocess have deadletter jobs.
So far AbortProcessAction do not wait and delete the process
hierarchy which lead to deadlock in DB.

Jira: LMCROSSITXSADEPLOY-1761

### CODE REVIEW

File: com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/flowable/FlowableFacade.java [hunk 1]
```diff
@@ -113,6 +113,11 @@ public boolean hasDeadLetterJobs(String processId) {
         return !getDeadLetterJobs(processId).isEmpty();
     }
 
+    private boolean haveAllLeafsHaveDeadLetterJobs(List<Execution> executions) {
```
Comment (enchobelezirev) [inline com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/flowable/FlowableFacade.java]: For me an you leaf is an understandable word but not for the rest of the team.
Maybe renaming it to child processes or subProcessLeafs would be better.

File: com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/flowable/FlowableFacade.java [hunk 2]
```diff
@@ -268,6 +280,25 @@ public boolean isProcessInstanceAtReceiveTask(String processInstanceId) {
                                    .collect(Collectors.toList());
     }
 
+    private List<Execution> getAllLeafExecutions(String processInstanceId) {
+        List<Execution> allExecutions = getAllProcessExecutions(processInstanceId);
+        return allExecutions.stream()
```
Comment (enchobelezirev) [inline com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/flowable/FlowableFacade.java]: This is a lot complexity: O(n^2).
Could we revert the things like this:
1. Find the execution which is related to the processInstanceId
2. Find each executions which have parentId = processInstaceId || superExecutionId = processInstanceId
3. Do 1 and 2, recursively for each execution foud in the step 2 
Would this work?
This way we would lower the complexity to O(nlogn)

File: com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/flowable/FlowableFacade.java [hunk 3]
```diff
@@ -230,6 +235,13 @@ public void deleteProcessInstance(String processInstanceId, String deleteReason)
                 // different if the process has parallel executions.
                 processEngine.getRuntimeService()
                              .setVariable(processInstanceId, Constants.PROCESS_ABORTED, Boolean.TRUE);
+                while (true) {
+                    List<Execution> allLeafExecutions = getAllLeafExecutions(processInstanceId);
+
+                    if (haveAllLeafsHaveDeadLetterJobs(allLeafExecutions)) {
+                        break;
+                    }
```
Comment (enchobelezirev) [inline com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/flowable/FlowableFacade.java]: What if there is a bug in the code? We would stay in the loop forever.
Maybe giving it some time would be enough in order the executions to finish. For instance, 10 mins MAX or more?

File: com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/flowable/FlowableFacade.java (hunk 1 above)
Comment (boyan-velinov) [reply com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/flowable/FlowableFacade.java]: Done

File: com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/flowable/FlowableFacade.java (hunk 3 above)
Comment (boyan-velinov) [reply com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/flowable/FlowableFacade.java]: Thanks for the idea!
Done

File: com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/flowable/FlowableFacade.java (hunk 2 above)
Comment (boyan-velinov) [reply com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/flowable/FlowableFacade.java]: About the complexity I do agree.
However, I am not sure if it worths to have recursion here in order to optimize a little bit when the code will not look nicer.

**Pull request merged**